virtcontainers/clh.go

```diff
@@ -0,0 +1,1260 @@
+	{"page_alloc.shuffle", "1"}, // enable page allocation shuffling
+}
+
+var clhDebugKernelParams = []Param{
+
+	{"console", "ttyS0,115200n8"},     // enable serial console
+	{"systemd.log_level", "debug"},    // enable systemd debug output
+    {"systemd.log_target", "console"}, // send loggng to the console
+	{"initcall_debug", "1"},           // print init call timing information to the console
+}
+
+//###########################################################
+//
+// hypervisor interface implementation for cloud-hypervisor
+//
+//###########################################################
+
+func (clh *cloudHypervisor) createSandbox(ctx context.Context, id string, networkNS NetworkNamespace, hypervisorConfig *HypervisorConfig, vcStore *store.VCStore) error {
+	clh.ctx = ctx
+
+	span, _ := clh.trace("createSandbox")
+	defer span.Finish()
+
+	err := hypervisorConfig.valid()
+	if err != nil {
+		return err
+	}
+
+	clh.id = id
+	clh.store = vcStore
+	clh.config = *hypervisorConfig
+	clh.state.state = clhNotReady
+
+	err = clh.getAvailableVersion()
+	if err != nil {
+		return err
+	}
+
+	if clh.version.Major < supportedMajorVersion && clh.version.Minor < supportedMinorVersion {
+		errorMessage := fmt.Sprintf("Unsupported version: cloud-hypervisor %d.%d not supported by this driver version (%d.%d)",
+			clh.version.Major,
+			clh.version.Minor,
+			supportedMajorVersion,
+			supportedMinorVersion)
+		return errors.New(errorMessage)
+	}
+
+	clh.cliBuilder = &DefaultCLIBuilder{}
+
+	socketPath, err := clh.vsockSocketPath(id)
+	if err != nil {
+		clh.Logger().Info("Invalid socket path for cloud-hypervisor")
+		return nil
+	}
+	clh.socketPath = socketPath
+
+	clh.Logger().WithField("function", "createSandbox").Info("creating Sandbox")
+
+	// No need to return an error from there since there might be nothing
+	// to fetch if this is the first time the hypervisor is created.
+	if err := clh.store.Load(store.Hypervisor, &clh.state); err != nil {
+		clh.Logger().WithField("function", "createSandbox").WithError(err).Info("No info could be fetched")
+	}
+
+	// Set initial memomory size of the cloud hypervisor
+	clh.cliBuilder.SetMemory(&CLIMemory{
+		memorySize:  clh.config.MemorySize,
+		backingFile: "/dev/shm",
+	})
+	// Set initial amount of cpu's for the cloud hypervisor
+	clh.cliBuilder.SetCpus(&CLICpus{
+		cpus: clh.config.NumVCPUs,
+	})
+
+	// Add the kernel path
+	kernelPath, err := clh.config.KernelAssetPath()
+	if err != nil {
+		return err
+	}
+	clh.cliBuilder.SetKernel(&CLIKernel{
+		path: kernelPath,
+	})
+
+	// First take the default parameters defined by this driver
+	clh.cliBuilder.AddKernelParameters(clhKernelParams)
+
+	// Followed by extra debug parameters if debug enabled in configuration file
+	if clh.config.Debug == true {
+		clh.cliBuilder.AddKernelParameters(clhDebugKernelParams)
+	}
+
+	// Followed by extra debug parameters defined in the configuration file
+	clh.cliBuilder.AddKernelParameters(clh.config.KernelParams)
```

Comment:
Maybe this could be grouped into a single `func (d *DefaultCLIBuilder) AddKernelParameters(hypervisorConfig *HypervisorConfig)` routine?